compiler/src/iree/compiler/Codegen/Dialect/GPU/Transforms/ConcretizeMmaShapes.cpp

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUEnums.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Passes.h"
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Transforms.h"
+#include "iree/compiler/Codegen/Dialect/VectorExt/IR/VectorExtDialect.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::IREE::GPU {
+
+#define GEN_PASS_DEF_CONCRETIZEMMASHAPESPASS
+#include "iree/compiler/Codegen/Dialect/GPU/Transforms/Passes.h.inc"
+
+namespace {
+struct ConcretizeMmaShapesPass final
+    : impl::ConcretizeMmaShapesPassBase<ConcretizeMmaShapesPass> {
+  using ConcretizeMmaShapesPassBase::ConcretizeMmaShapesPassBase;
+  void runOnOperation() override;
+};
+} // namespace
+
+struct ConcretizeMmaInputShapes final
+    : OpRewritePattern<IREE::GPU::MultiMmaOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::GPU::MultiMmaOp mmaOp,
+                                PatternRewriter &rewriter) const override {
+    IREE::GPU::MmaInterfaceAttr kind = mmaOp.getKind();
+    SmallVector<ReassociationIndices> lhsReassociations, rhsReassociations;
+    RankedTensorType lhsConcreteType, rhsConcreteType;
+    if (failed(kind.materializeOperandConcreteShape(
+            rewriter, IREE::GPU::MMAFragment::Lhs, mmaOp.getLhs(),
+            mmaOp.getLhsPermutation(), lhsReassociations, lhsConcreteType))) {
+      return failure();
+    }
+    if (failed(kind.materializeOperandConcreteShape(
+            rewriter, IREE::GPU::MMAFragment::Rhs, mmaOp.getRhs(),
+            mmaOp.getRhsPermutation(), rhsReassociations, rhsConcreteType))) {
+      return failure();
+    }
+    Location loc = mmaOp->getLoc();
+    Value concreteLhs =
+        rewriter
+            .create<tensor::ExpandShapeOp>(loc, lhsConcreteType, mmaOp.getLhs(),
+                                           lhsReassociations)
+            .getResult();
+    Value concreteRhs =
+        rewriter
+            .create<tensor::ExpandShapeOp>(loc, rhsConcreteType, mmaOp.getRhs(),
+                                           rhsReassociations)
+            .getResult();
+
+    std::optional<DenseI64ArrayAttr> lhsPerm, rhsPerm, accPerm;
+    if (mmaOp.getLhsPermutation().has_value())
+      lhsPerm = mmaOp.getLhsPermutationAttr();
+    if (mmaOp.getRhsPermutation().has_value())
+      rhsPerm = mmaOp.getRhsPermutationAttr();
+    if (mmaOp.getAccPermutation().has_value())
+      accPerm = mmaOp.getAccPermutationAttr();
+
+    rewriter.replaceOpWithNewOp<IREE::GPU::MultiMmaOp>(
+        mmaOp, concreteLhs, concreteRhs, mmaOp.getAcc(),
+        mmaOp.getIndexingMaps(), mmaOp.getIteratorTypes(), mmaOp.getKind(),
+        lhsPerm, rhsPerm, accPerm);
```

Comment:
Preserve lowering config on the op if present (getLoweringConfig, setLoweringConfig)